From 8d588ea1ce684aed8a1304ca1006d98758ec3195 Mon Sep 17 00:00:00 2001
From: Hans-Kristian Arntzen <post@arntzen-software.no>
Date: Wed, 26 Aug 2026 13:15:47 +0200
Subject: [PATCH 06/19] vkd3d: Warn about broken out of band transition cases.

Signed-off-by: Hans-Kristian Arntzen <post@arntzen-software.no>
---
 libs/vkd3d/command_queue_vkd3d_ext.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/libs/vkd3d/command_queue_vkd3d_ext.c b/libs/vkd3d/command_queue_vkd3d_ext.c
index 95a96172..5e12740c 100644
--- a/libs/vkd3d/command_queue_vkd3d_ext.c
+++ b/libs/vkd3d/command_queue_vkd3d_ext.c
@@ -75,6 +75,9 @@ static HRESULT STDMETHODCALLTYPE d3d12_command_queue_vkd3d_ext_NotifyOutOfBandCo
         {
             struct vkd3d_queue *out_of_band_queue = command_queue->device->queue_families[i]->out_of_band_queue;
 
+            if (command_queue->last_submission_timeline_value != 0)
+                FIXME("Trying to assign a queue to be out of band, but it has been submitted to before. This may not work as intended.\n");
+
             /* Migrate the command queue to the out of band queue, so that it cannot conflict
              * with other queues from scheduling PoV.
              * The assumption is that NvAPI notifies before the queue has actually been used
